rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -151,8 +178,15 @@ impl MultisigCheckpointSyncer {
         let mut signed_checkpoints_per_root: HashMap<H256, Vec<SignedCheckpointWithMessageId>> =
             HashMap::new();
 
-        for validator in validators.iter() {
-            let addr = H160::from(*validator);
+        let sorted_validators: Vec<(usize, &ValidatorWithWeight)> = weighted_validators
+            .iter()
+            .enumerate()
+            .sorted_by_key(|(_, vw)| std::cmp::Reverse(vw.weight))
+            .collect();
+
+        let mut cumulative_weight: u64 = 0;
```

Comment:
we shouldn't have a cumulative_weight defined at this level

the logic in this function is intended to be robust to validators in a set signing different roots, in which case it shouldn't be considered a quorum. That's why we have this `signed_checkpoints_per_root` HashMap keyed by roots. If everything works correctly, there should only ever be a single key, but this makes us tolerant to misbehaving validators.

Now seeing this we should probably extend this key to not be the merkle root but instead the digest of the checkpoint, as it's possible for misbehaving validators to modify other things like the merkle tree hook address / domain / message id. So instead of `let root = signed_checkpoint.value.root;`, something like `let root = signed_checkpoint.value.signing_hash();`

But anyways imo this should match the logic we had before, where instead of having a mutable `signature_count` that would be added to with each found checkpoint, we only look at the cumulatively found weights in the appropriate entry of `signed_checkpoints_per_root`

Reply:
https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/4341 is related I think